### ISSUE TITLE
Improve documentation

### ISSUE THREAD

**Issue by auvipy** [mirumee/saleor#357]
I didn't find any proper documentation of the project.

**Comment by rogueleaderr** [comment]
+1 on this. I'm evaluating eCommerce options and I'm tempted to choose Saleor but I just can't jump in with such limited documentation.

**Comment by artursmet** [comment]
What kind of documentation would be the best for you? Some developer docs, like "where should I put my custom code, to achieve it", or something like out of the box feature list with short description?

**Comment by awhillas** [comment]
I find example code the most helpful, so the first option I guess. 

There are 2 use cases for product design that stand out in my mind:
- A single complex product with many attributes/variations i.e. each attribute combination gives a different price/image/SKUs etc
- Many simple products each with their own set of attributes i.e. a department store with thousands of different products i.e. too many to create individual class definitions for so mass importing is desirable.

Also, complex pricing is always difficult to model i.e. different prices for different user groups, taxes for different countries, multiple currencies, discounts, sales etc.

Product attributes/variants and pricing structure are the two most challenging parts of shopping cart design i have found.

**Comment by rogueleaderr** [comment]
In my mind, the four key elements of good documentation are:

1) A conceptual overview. Many users who come to Saleor will have never used an eCommerce system before. The docs should clearly explain (in plain English) all of the key things the framework does. Then it should clearly lay out what the key pieces are and what they do.
2) Example code / tutorials. Users who have standard use cases should by-and-large be able to get up and running just by using default code and following a tutorial.
3) A recipe book: clear instructions on how to make the most common customizations
4) API docs laying out the classes and explaining what they do and how to use them

I'm currently evaluating eCommerce frameworks and so I'm looking at Django Oscar which has very clear and [extensive documentation](https://django-oscar.readthedocs.org/en/releases-1.1/). I'm not sure if it's actually better for my use case but the documentation makes me feel like at least I'll be able to understand what's going on while I figure it out (without having to read a mountain of source code).

**Comment by awhillas** [comment]
Yeah, I was looking at Django Oscar but the code is a nightmare:

- weird "forking" which is what they call inheritance overriding of their basic apps. Lots of non-standard Django stuff and the payoff doesn't seem to be there?
- Products are defined in the Catalogue app which means you can't leave the Oscar catalogue organisation behind and come up with your own.
- Prices are grouped into Strategy's (good but...) which couple stock and pricing information (bad for digital products) and are, of course, kept in the Partner app (and Products are in the Catalogue app?)?

and a bunch of other things (attributes + options which appear to be the same thing, you get a ton of apps when you install half of which you don't need). Documentation is more but dated and filled with many holes and light on examples. 

Saleor seems simpler and they have broken their design down into separate projects (satchless, django-payments etc) so coupling is nice and lose and modular. Seems like design is first.

**Comment by awhillas** [comment]
Yeah, I was looking at Django Oscar but the code is a nightmare:

- weird "forking" which is what they call inheritance overriding of their basic apps. Lots of non-standard Django stuff and the payoff doesn't seem to be there?
- Products are defined in the Catalogue app which means you can't leave the Oscar catalogue organisation behind and come up with your own.
- Prices are grouped into Strategy's (good but...) which couple stock and pricing information (bad for digital products) and are, of course, kept in the Partner app (and Products are in the Catalogue app?)?

and a bunch of other things (attributes + options which appear to be the same thing, you get a ton of apps when you install half of which you don't need). Documentation is more but dated and filled with many holes and light on examples. 

Saleor seems simpler and they have broken their design down into separate projects (satchless, django-payments etc) so coupling is nice and lose and modular. Seems like design is first.

**Comment by auvipy** [comment]
django project itself is a standard for doc and drf

**Comment by awhillas** [comment]
We are not talking tech standards, we're talking content

**Comment by auvipy** [comment]
I wan intending to point to the contents of django docs actually

**Comment by artursmet** [comment]
So I suggest to focus on following steps:
- [ ] Create basic description of saleor project concepts
- [ ] Create out of the box features list
- [ ] Improve code documentation (create docstrings for important classes and methods)
- [ ] Collect some common recipes in one place

**Comment by artursmet** [comment]
Initial version of project documentation was deployed on readthedocs: http://saleor.readthedocs.org/en/latest/

For now, it's only content from project's wiki, but I think it's a good start for documentation.

**Comment by pcompassion** [comment]
I expected to install saleor as an `app` to my existing django project. 

Upon reading Install instruction, I suspect it's not the case and saleor runs as a separate server.  

Am I correct on this? 

Is there a way to integrate saleor into an existing django project?

**Comment by patrys** [comment]
Saleor is a project template that is meant to be customized for your needs. You can integrate it into your project by copying the `saleor` app to your project directory. Just make sure your settings file has all of the relevant values defined.

**Comment by pcompassion** [comment]
I'm new to whole e-commerce concept, so bear with me please.

A bit of explanation on difference among `product - product_variant - product_attribute - attribute_choice_value` would be helpful.  

There's also ProductVariant's `attributes`


product: starwars episode1
product_variant: paperbook /kindle

product_attributes: umm? shouldn't this be product_variant_attributes?

I guess the structure could change based on intention. 
It would be helpful to see an intended usage example though. 
I guess it could be on the `random_data.py` or documentation.

Attribute can have many AttributeChoiceValue and I don't understand the reasoning behind this either.

**Comment by artursmet** [comment]
You have products, and product variants, so your product could be `Star Wars Episode 1` and it will have `format` attribute.
Now you have two ProductVariants with different `format` attribute values `paperbook` and `kindle`.

Attributes in product are possible options for variants (format, color, size etc) and `AttributeChoiceValue` are values for particular variants (paperbook, red, XXL, etc)

**Comment by patrys** [comment]
Like @artursmet says, `product_attributes` let you decide what kinds of attributes that particular product's variants can have. If you sell shoes and books you don't want your staff to wonder how to calculate number of pages for a pair of sneakers or to enter "size 36" for a book. :wink: Hope this helps.

**Comment by pcompassion** [comment]
Thank you. helps immensely

So ProductVariant's attributes(jsonfield) indicates what options(product_attributes) are available for it.

`paperbook` and `kindle` would be AttributeChoiceValue 's `display

and AttributeChoiceValue's `color` and `image` are kind of a meta field (it doesn't mean color of a product)

**Comment by artursmet** [comment]
`color` and `image` fields were meant to improve displaying particular attributes on product page. For example when attached image to attribute value, you can display it on variant selector on product page.
When you set color (hex value), you can display colored box on product page.

**Comment by pcompassion** [comment]
yes. Thank you. cleared all my questions except one. 

Am I correct on 

`ProductVariant's attributes(jsonfield) indicates what options(product_attributes) it chooses to support.`

@patrys ' comment 
`product_attributes let you decide what kinds of attributes that particular product's variants can have.` 

confuses me. 

It's the `ProductVariant.attributes` which let you decide  what kinds of attributes that particular product's variants can have. right?


Actually another question is..

There's no ordering of `variant` and `attribute`, right?

i.e. If I wanted to, I could let user to select variant first and then attribute, and vice-versa?

@artursmet 's sentence `you can display it on variant selector on product page.` suggest the idea. (I'm trying to understand the sentence)

**Comment by patrys** [comment]
Closing this in favour of #157
